apps/rush-lib/src/cli/actions/AddAction.ts

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as os from 'os';
+
+import {
+  CommandLineFlagParameter,
+  CommandLineStringParameter
+} from '@microsoft/ts-command-line';
+
+import { RushConfigurationProject } from '../../api/RushConfigurationProject';
+import { BaseRushAction } from './BaseRushAction';
+import { RushCommandLineParser } from '../RushCommandLineParser';
+import { DependencyIntegrator, SemVerStyle } from '../../logic/DependencyIntegrator';
+
+export class AddAction extends BaseRushAction {
+  private _exactFlag: CommandLineFlagParameter;
+  private _caretFlag: CommandLineFlagParameter;
+  private _devDependencyFlag: CommandLineFlagParameter;
+  private _makeConsistentFlag: CommandLineFlagParameter;
+  private _noInstallFlag: CommandLineFlagParameter;
+  private _packageName: CommandLineStringParameter;
+  private _versionSpecifier: CommandLineStringParameter;
+
+  constructor(parser: RushCommandLineParser) {
+    const documentation: string[] = [
+      'Blah.'
+    ];
+    super({
+      actionName: 'add',
+      summary: 'Adds a dependency to the package.json and runs rush upgrade.',
+      documentation: documentation.join(os.EOL),
+      safeForSimultaneousRushProcesses: false,
+      parser
+    });
+  }
+
+  public onDefineParameters(): void {
+    this._packageName = this.defineStringParameter({
+      parameterLongName: '--package',
+      parameterShortName: '-p',
+      required: true,
+      argumentName: 'PACKAGE_NAME',
+      description: '(Required) The name of the package which should be added as a dependency'
+    });
+    this._versionSpecifier = this.defineStringParameter({
+      parameterLongName: '--version',
+      parameterShortName: '-v',
+      argumentName: 'VERSION_RANGE',
+      description: ''
+    });
+    this._exactFlag = this.defineFlagParameter({
+      parameterLongName: '--exact',
+      description: 'If specified, the version specifier inserted into the'
+        + ' package.json will be a locked, exact version.'
+    });
+    this._caretFlag = this.defineFlagParameter({
+      parameterLongName: '--caret',
+      description: 'If specified, the version specifier inserted into the'
+        + ' package.json will be a prepended with a "caret" specifier ("^").'
+    });
+    this._devDependencyFlag = this.defineFlagParameter({
```

Comment:
These can be just specified with the version.

Reply:
?

---
In reply to: [220407415](https://github.com/Microsoft/web-build-tools/pull/843#discussion_r220407415) [](ancestors = 220407415)

Reply:
You can specify the version as "~1.2.3" or "^2.3.4"

---
In reply to: [220743430](https://github.com/Microsoft/web-build-tools/pull/843#discussion_r220743430) [](ancestors = 220743430,220407415)

Reply:
You highlighted devDependency flag.

And that isn't how --version works. It works how "npm install --save" works.

---
In reply to: [220750046](https://github.com/Microsoft/web-build-tools/pull/843#discussion_r220750046) [](ancestors = 220750046,220743430,220407415)

Reply:
So NPM has --carat and --tilde arguments? That seems really clunky.

---
In reply to: [220751772](https://github.com/Microsoft/web-build-tools/pull/843#discussion_r220751772) [](ancestors = 220751772,220750046,220743430,220407415)